internal/controller/postgrescluster/autogrow.go

```diff
@@ -0,0 +1,189 @@
+	limitSet := limitIsSet(cluster, volumeType, instanceSetName)
+
+	if limitSet && current.Value() > previous.Value() {
+		r.Recorder.Eventf(cluster, corev1.EventTypeNormal, "VolumeAutoGrow",
+			"%s volume expansion to %v requested for %s/%s.",
+			volumeType, current.String(), cluster.Name, instanceSetName)
+	}
+
+	// If the desired size was not observed, update with previously stored value.
+	// This can happen in scenarios where the annotation on the Pod is missing
+	// such as when the cluster is shutdown or a Pod is in the middle of a restart.
+	if desiredRequest == "" {
+		desiredRequest = desiredRequestBackup
+	}
+
+	return desiredRequest
+}
+
+// limitIsSet determines if the limit is set for a given volume type and returns
+// a corresponding boolean value
+func limitIsSet(cluster *v1beta1.PostgresCluster, volumeType, instanceSetName string) bool {
+
+	var limitSet bool
+
+	switch volumeType {
+
+	// Cycle through the instance sets to ensure the correct limit is identified.
+	case "pgData":
+		for _, specInstance := range cluster.Spec.InstanceSets {
+			if specInstance.Name == instanceSetName {
+				limitSet = !specInstance.DataVolumeClaimSpec.Resources.Limits.Storage().IsZero()
+			}
+		}
+	}
+	// TODO: Add cases for pgWAL and repo volumes
+
+	return limitSet
+
+}
+
+// setVolumeSize compares the potential sizes from the instance spec, status
+// and limit and sets the appropriate current value.
+func (r *Reconciler) setVolumeSize(ctx context.Context, cluster *v1beta1.PostgresCluster,
+	pvc *corev1.PersistentVolumeClaim, volumeType, instanceSpecName string) {
+
+	log := logging.FromContext(ctx)
+
+	// Store the limit for this instance set. This value will not change below.
+	volumeLimitFromSpec := pvc.Spec.Resources.Limits.Storage()
+
+	// Capture the largest volume size currently defined for a given instance set.
+	// This value will capture our desired update.
+	volumeRequestSize := pvc.Spec.Resources.Requests.Storage()
+
+	// A limit of 0 is ignorned, so the volume request is used.
+	if volumeLimitFromSpec.IsZero() {
+		return
+	}
+
+	// If the request value is greater than the set limit, use the limit and issue
+	// a warning event.
+	if volumeRequestSize.Value() > volumeLimitFromSpec.Value() {
+		r.Recorder.Eventf(cluster, corev1.EventTypeWarning, "VolumeRequestOverLimit",
+			"%s volume request (%v) for %s/%s is greater than set limit (%v). Limit value will be used.",
+			volumeType, volumeRequestSize, cluster.Name, instanceSpecName, volumeLimitFromSpec)
+
+		pvc.Spec.Resources.Requests = corev1.ResourceList{
+			corev1.ResourceStorage: *resource.NewQuantity(volumeLimitFromSpec.Value(), resource.BinarySI),
+		}
+		// Otherwise, if the limit is not set or the feature gate is not enabled, do not autogrow.
+	} else if feature.Enabled(ctx, feature.AutoGrowVolumes) {
+
+		// determine the appropriate volume request based on what's set in the status
+		if dpv, err := getDesiredVolumeSize(
```

Comment:
what do we think the "P" in `dpv` means?

Reply:
Oh is see, `volumeRequestSize` is a pointer that is passed in and updated. 

it took me a minute to get here - maybe its worth changing something here ♻️ 

- if we aren't using the return value maybe we don't return it
- or maybe return the value instead of updating the pointer

Reply:
If we simplified to that logic, I think that eliminates the warning event for a defined request that is higher than the limit. That's a useful warning in my opinion because it helps explains unexpected behavior (due to a misconfiguration on the user's part; you shouldn't even define a request that's higher than the limit you define in the same object).

Reply:
For the return by reference, we could definitely return it explicitly. I'm not returning anything that's not used, it's just that the one value is only used in the error log, which feels ok to me.

Reply:
It would eliminate the `VolumeRequestOverLimit` Event if I'm following correctly.

I was trying to keep the pulled out function simple, which also makes the test less complicated. It seemed better to just return the string in that case. :man_shrugging: